crates/ty_python_semantic/src/types.rs

```diff
@@ -2468,6 +2468,34 @@ impl<'db> Type<'db> {
                 })
             }
 
+            // TODO: The following two branches are a temporary workaround to prevent an important union type
+            // in numpy's internals from collapsing to a single type, which causes thousands of downstream
+            // false positive diagnostics across the ecosytem. See https://github.com/astral-sh/ty/issues/1666
+            // for details.
+            (Type::SubclassOf(self_subclass_ty), Type::ProtocolInstance(_))
+                if self_subclass_ty.is_type_var() =>
```

Comment:
but `type[T]` should always be _assignable_ to a protocol-instance if `T`'s upper bound is assignable to the protocol-instance, right? 

```suggestion
            (Type::SubclassOf(self_subclass_ty), Type::ProtocolInstance(_))
                if self_subclass_ty.is_type_var() && relation.is_assignability() =>
```

Reply:
Do you mean: this branch should *return* `ConstraintSet::from(relation.is_assignability())`? Because if we guard it with `is_assignability`, the workaround would not be effective.

Reply:
Ah, right, yeah -- I think I got the polarity the wrong way round?

```suggestion
            (Type::SubclassOf(self_subclass_ty), Type::ProtocolInstance(_))
                if self_subclass_ty.is_type_var() && !relation.is_assignability() =>
```

I don't think the body of the branch should be `ConstraintSet::from(relation.is_assignability())` -- if the relation is `TypeRelation::Assignability`, we only want the assignability check to hold for `type[T] <: P` if the meta-type of `T`'s upper bound is assignable to `P` (we want to fall through to a later branch that should handle that for us)